src/main/kotlin/com/terraformation/backend/eventlog/db/EventLogStore.kt

```diff
@@ -0,0 +1,208 @@
+          requestedConcreteClasses.map {
+            EVENT_CLASS.like(it.java.name.substringBeforeLast('V') + "V%")
+          }
+      val condition =
+          DSL.and(
+              payloadField(idFieldName, idDataType).eq(id),
+              DSL.or(likeConditions),
+          )
+
+      fetchByCondition(condition, requestedConcreteClasses)
+    }
+  }
+
+  private fun <T : PersistentEvent> fetchByCondition(
+      condition: Condition,
+      requestedClasses: Collection<KClass<out T>>,
+  ): List<EventLogEntry<T>> {
+    requestedClasses.forEach { requestedClass ->
+      if (requestedClass.isSubclassOf(UpgradableEvent::class)) {
+        throw IllegalArgumentException(
+            "${requestedClass.java.name} is upgradable; only the latest version may be queried"
+        )
+      }
+      if (requestedClass.isAbstract || requestedClass.java.isInterface) {
+        throw IllegalArgumentException("${requestedClass.java.name} is not a concrete class")
+      }
+    }
+
+    return with(EVENT_LOG) {
+      dslContext
+          .select(CREATED_BY, CREATED_TIME, EVENT_CLASS, ID, PAYLOAD)
+          .from(EVENT_LOG)
+          .where(condition)
+          .orderBy(CREATED_TIME, ID)
+          .fetch { eventLogEntryFromRecord(it, requestedClasses) }
+    }
+  }
+
+  /**
+   * Returns an [EventLogEntry] containing the latest version of an event from an [EVENT_LOG]
+   * record. If the event as stored in the database is using an out-of-date version of the event
+   * class, updates the database to store the new version.
+   */
+  private fun <T : PersistentEvent> eventLogEntryFromRecord(
+      record: Record,
+      requestedClasses: Collection<KClass<out T>>,
+  ): EventLogEntry<T> {
+    val className = record[EVENT_LOG.EVENT_CLASS]!!
+    val rawEventObject =
+        objectMapper.readValue(record[EVENT_LOG.PAYLOAD]!!.data(), Class.forName(className))
+    if (rawEventObject !is PersistentEvent) {
+      throw IllegalStateException("Event class $className does not implement PersistentEvent")
+    }
+
+    val eventObject = upgradeToLatest(rawEventObject)
+
+    if (requestedClasses.none { it.isInstance(eventObject) }) {
+      val requestedClassNames = requestedClasses.joinToString()
+      throw IllegalStateException(
+          "Event class $className is not on requested list: $requestedClassNames"
+      )
+    }
+
+    if (eventObject != rawEventObject) {
+      writeNewEventVersion(record[EVENT_LOG.ID]!!, eventObject)
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    return EventLogEntry(
+        createdBy = record[EVENT_LOG.CREATED_BY]!!,
+        createdTime = record[EVENT_LOG.CREATED_TIME]!!,
+        event = eventObject as T,
+        id = record[EVENT_LOG.ID]!!,
+    )
+  }
+
+  /**
+   * Updates an event's payload and class in the database after it is upgraded to a new version. The
+   * original class and payload are preserved.
+   */
+  private fun writeNewEventVersion(eventLogId: EventLogId, eventObject: PersistentEvent) {
```

Comment:
Does this mean that we only store the previous event's version's payload and class? (not saying we need to store more, just checking my understanding)

Reply:
Updated the comment to hopefully make this clearer.